components/raftstore/src/store/peer_storage.rs

```diff
@@ -400,6 +400,38 @@ pub fn init_apply_state(engines: &Engines, region: &Region) -> Result<RaftApplyS
     )
 }
 
+fn validate_states(raft_state: &RaftLocalState, apply_state: &RaftApplyState) -> Result<()> {
+    let last_index = raft_state.get_last_index();
+    let commit_index = raft_state.get_hard_state().get_commit();
+    let apply_index = apply_state.get_applied_index();
+    if last_index < commit_index {
+        return Err(box_err!("raft_state {:?} corrupted", raft_state));
+    }
+
+    if commit_index >= apply_index {
+        if commit_index >= apply_state.get_last_commit_index() {
+            return Ok(());
+        }
+        return Err(box_err!(
+            "raft log lost detected: {:?} {:?}",
+            raft_state,
+            apply_state
+        ));
+    }
```

Comment:
Is it necessary to judge where `commit_index` is larger than `last_commit_index` ? We would judge it at next statement.

Reply:
I don't get it. Do you mean not to check L412? The check is to detect data lost even the final result still meet all the other constraint.